File: sentil-matlab/+sentil/LiftingRegistry.m
classdef LiftingRegistry < handle
    %LIFTINGREGISTRY Per-variable noise models used to lift a clean trace.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods
        function obj = LiftingRegistry(handle)
            if nargin == 1 && isa(handle, 'uint64') && isscalar(handle)
                obj.Handle = handle;
            else
                obj.Handle = sentil_mex('lifting_create');
            end
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('lifting_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function register(obj, variable, model, interaction)
            %REGISTER Attach a noise model to a variable. The model is consumed.
            obj.assertOpen();
            if nargin < 4
                interaction = sentil.NoiseInteraction.Additive;
            end
            registry = obj.Handle;
            sentil_mex('lifting_register', registry, char(variable), model.consume(), ...
                double(int32(interaction)));
        end

        function v = variables(obj)
            %VARIABLES The variables that carry a noise model, sorted.
            obj.assertOpen();
            v = sentil_mex('lifting_variables', obj.Handle);
        end

        function tf = is_empty(obj)
            %IS_EMPTY Whether no variable carries a noise model.
            obj.assertOpen();
            tf = sentil_mex('lifting_is_empty', obj.Handle);
        end

        function t = lift(obj, trace, seed)
            %LIFT One seeded noisy realization of the trace.
            obj.assertOpen();
            t = sentil.Trace(uint64(sentil_mex('lifting_lift', obj.Handle, trace.Handle, ...
                double(seed))));
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this registry has been closed');
            end
        end
    end
end

File: sentil-java/src/main/java/io/github/sedislab/sentil/LiftingRegistry.java
package io.github.sedislab.sentil;

import java.util.Arrays;
import java.util.List;

/** The per-variable noise models that lift a deterministic trace to a stochastic ensemble. */
public final class LiftingRegistry extends NativeResource {
    LiftingRegistry(long handle) {
        super(handle, NativeLib::liftingDestroy);
    }

    public LiftingRegistry() throws SentilException {
        this(NativeLib.liftingCreate());
    }

    /** Attach a noise model to a variable. The model is consumed. */
    public void register(String variable, NoiseModel model, NoiseInteraction interaction)
            throws SentilException {
        long registry = handle();
        long modelHandle = model.consume();
        NativeLib.liftingRegister(registry, variable, modelHandle, interaction.code());
    }

    /** Attach a noise model to a variable with the default additive interaction. */
    public void register(String variable, NoiseModel model) throws SentilException {
        register(variable, model, NoiseInteraction.ADDITIVE);
    }

    /** The variables that carry a noise model, sorted. */
    public List<String> variables() throws SentilException {
        return Arrays.asList(NativeLib.liftingVariables(handle()));
    }

    /** Whether no variable carries a noise model. */
    public boolean isEmpty() {
        return NativeLib.liftingIsEmpty(handle());
    }

    /** One seeded noisy realization of the trace. */
    public Trace lift(Trace trace, long seed) throws SentilException {
        return new Trace(NativeLib.liftingLift(handle(), trace.handle(), seed));
    }
}